### PR TITLE
Add backoff to the consul client

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes https://github.com/weaveworks/service/issues/423

### CODE REVIEW

**Comment by 2opremio** [comment]
Wouldn't it be cleaner to reuse `common/backoff` ?

**Comment by tomwilkie** [comment]
`common/backoff` is for periodic tasks, this one is continuous (ie not delay between iterations).  Also `common/backoff` logging behaviour is not appropriate here.

I considered changing common/backoff to be more general purpose, but I didn't think it was the right call.  I could be persuaded otherwise, but it will make common/backoff more complicated.

File: app/multitenant/consul_client.go [hunk 1]
```diff
@@ -133,21 +133,40 @@ func (c *consulClient) CAS(key string, out interface{}, f CASCallback) error {
...
+			}
+			select {
+			case <-done:
+				return
+			case <-time.After(backoff):
+				continue
+			}
+		}
+		backoff = initialBackoff
+		if index == meta.LastIndex {
```
Comment (2opremio) [inline app/multitenant/consul_client.go]: Duplicated if

File: app/multitenant/consul_client.go [hunk 2]
```diff
@@ -133,21 +133,40 @@ func (c *consulClient) CAS(key string, out interface{}, f CASCallback) error {
...
+		index   = uint64(0)
+	)
 	for {
 		select {
 		case <-done:
 			return
 		default:
 		}
-
 		kvps, meta, err := c.kv.List(prefix, &consul.QueryOptions{
```
Comment (2opremio) [inline app/multitenant/consul_client.go]: I was looking for a way to cancel the request directly but there doesn't seem to be a way, meh

**Comment by 2opremio** [comment]
> I considered changing common/backoff to be more general purpose, but I didn't think it was the right call.

Fair enough.

**Comment by 2opremio** [comment]
LGTM (comment-aside)

**Pull request merged**